### ISSUE TITLE
Create user email verification system

### ISSUE THREAD

**Issue by maxachis** [Police-Data-Accessibility-Project/data-sources-app#463]
### Context

* If we are going to create users based on the provided email, then we'll need to ensure that users are providing valid emails which they own and can access.
* Endpoints such as `/request-reset-password` or the eventual #26 require an extant email to operate properly.

#### Backend
* [x] Update standard user creation logic such that users must first confirm their email address with an email verification prior to becoming a full user.
* [x] Update the user creation endpoint to be similar to the `/reset-password` and `/request-reset-password` structure (`/auth/request-email-validation` and `/auth/signup-with-code`).
* [x] Consequently, the existing `/user` `POST` endpoint will need to be _🔥destroyed🔥_
* [x] This will likely require the creation of a new endpoint: `/signup-email-request` or something like that.
* [x] We may need to modify `/users` `POST`, as well. 
* [x] Add checks at schema and database level to ensure that emails present have valid structure (use the Marshmallow `Email` field in the app, but on the database level just check for the presence of an `@` -- we'll mostly validate emails with verifications)
* [x] Design this so that it's decoupled from the `/signup` logic: we may want people to be able to provide alternate emails for notifications in the future, and being able to re-used the same endpoint (or at least the same backend logic) for multiple purposes would be useful.
* [x] BCC josh.chamberlain@pdap.io on this email so they can follow up with people and generally have visibility on new signups
* [x] Add persistent users in database for testing

#### Frontend
* [x] Update frontend logic to handle new signup method. 

### Tests

* New tests will need to be created to test the new logic.
* Testing the mailgun component will be limited -- we'll need to mock that

### Docs

* API documentation will need updated

### Email template

from: support@pdap.io

```
Welcome to PDAP! If you meant to create an account, please verify your email by clicking this link: <link>

We're happy to have you in our community. If you'd like, you can reply to this email with a bit more about what you hope to accomplish, or if there's any way we can support your project. You are also invited to join us in Discord, where you can chat with our team and other folks interested in police data: https://discord.gg/wMqex8nKZJ
```

**Comment by maxachis** [comment]
@joshuagraber Sanity check requested! Do you elf eyes think this issue (and especially the API design discussed) seem kosher?

**Comment by josh-chamberlain** [comment]
@maxachis Since we allow github auth, I wonder if we should decouple this from login—we could let people specify a notification email (which might use their supplied one by default). There are various reasons people might want to use different email for login and notification anyway. One benefit may be that this is less entangled with auth, and just becomes a neat little property. It might also be more complicated, depending on how you look at it. Just a suggestion!

**Comment by maxachis** [comment]
@josh-chamberlain I think that's fair! This could be updated so that any email requires verification before it's entered into our database! Added a note about this in the requirements!

**Comment by joshuagraber** [comment]
Agreed with @josh-chamberlain. In other contexts where I've implemented similar functionality, contact verification is done independently of auth.

If you wanted to create a separate namespace, `/notification`, this could encapsulate all notification-related logic. (We don't have much for now, but I could see this being very useful down the line...) 

`/notification/contact/verify`, `POST`, sends verification email to email address passed as req data, and associates this as a notification contact for the user.
and `/notification/contact/verify/:token`, `POST` with token sent to email by the `/verify` endpoint.

Then `/notification/password-reset` `POST` can send the PW reset email with token that can then be passed to `/auth/reset-password`.

I get that a separate namespace may seem like overkill for now, since we'd only be using it for two emails, but...

...I believe we are angling toward eventually sending event-driven updates about user's followed searches, etc. With the namespace in place, we could add a templating system with handlebars, where a `POST` to, say,  `/notification/send` with `{ contact: <email/sms/device_id_for_push/whatev>, template: <template_name>, template_data: <data to interpolate into template>, ... } ` would send an email/SMS/push to the desired location, using the desired template, with the desired data.

Setting up the namespace now sets us up for that future success, without much in the way of premature optimization.

**Comment by josh-chamberlain** [comment]
All this makes sense to me! Notification is something we'd like to be very professional at, since it's one of the only features we offer

**Comment by maxachis** [comment]
@joshuagraber @josh-chamberlain So here is my thinking for the initial signup flow, and I want to get y'all's input on it:

1. `/user/signup` to perform initial signup (on browser only). Returns user id. User exists in the system but is not verified -- attempts at login will inform user they need to verify their email address
2. `/notification/contact/{user_id}/verify` `POST` sends verification email to user based on their email provided. The json body can mention that this is verification for the primary email, to properly direct the logic in 3 and allow the option for secondary emails.
3. `/notification/contact/{user_id}/verify/{verification_token}` `POST` confirms email. User is now verified. 

There are now some other things which need to be taken into account:
* If a user signs up, but hasn't verified their email address, how to handle `/login-with-github`? My thinking is to overwrite a pending primary email verification if they perform `/login-with-github` before verifying their email.
* `/link-with-github` should still be fine, because that will require an access token, which they can only get if they login successfully.

**Comment by joshuagraber** [comment]
> /user/signup to perform initial signup (on browser only). Returns user id. User exists in the system but is not verified -- attempts at login will inform user they need to verify their email address

Do you think we should block them from accessing the app entirely? Why not let them in, but disable certain functionality until they verify? We can add an annoying little banner in the profile to prompt them to do it, and have messaging around the disabled functionality.

> If a user signs up, but hasn't verified their email address, how to handle /login-with-github? My thinking is to overwrite a pending primary email verification if they perform /login-with-github before verifying their email.

If we allow them into the app but disable certain functionality, logging in and linking their account with github could work the same way. The messaging could just be "Verify email or link your account with Github to access this feature" or whatever.

Wyt?

**Comment by maxachis** [comment]
@joshuagraber Those seem like viable options to me, although I want to check with @josh-chamberlain to see their thoughts on what seems most viable.

**Comment by josh-chamberlain** [comment]
@joshuagraber @maxachis I co-sign Joshua's ideas.

I think we only need to verify email if they add one outside of the github signup—github should be handling verification on their end. Right?

As far as access to functionality—if people are not logged in, the nicest thing to do is to give them an in-context warning that they need to sign in to access whatever they're trying to do, and even capture their original intent so that when they sign up they are at the page they wanted (profile, making a request, whatever); that said, I think it's a fine default to simply take people to the sign in page if they try to do something which requires it.

**Comment by maxachis** [comment]
@joshuagraber @josh-chamberlain 
> If we allow them into the app but disable certain functionality, 

The simplest way to do this, from my end, is that anything which requires a JWT or Access Token cannot be done. 

If the user tries to log in but haven't verified, we say "Ahoy, matey, you have to verify your email before you can log in. 🦜"

Alternatively, we could have it so that it only occurs for JWT-required actions, or get more granular, but this is simplest for me personally, because it limits the amount of backend logic I need to change. In addition to modifying the user signup logic, I need to just modify the `/login` logic.

What do y'all think?

**Comment by joshuagraber** [comment]
> The simplest way to do this, from my end, is that anything which requires a JWT or Access Token cannot be done.

Why couldn't we just encode this into the access token?

```json
{
  ...,
  "verification": {
    "email": false
  }
}
```

That way the client can read it and show messaging around verifying, while also not blocking the user from accessing parts of the app that do not require verification.

**Comment by maxachis** [comment]
> > The simplest way to do this, from my end, is that anything which requires a JWT or Access Token cannot be done.
> 
> Why couldn't we just encode this into the access token?
> 
> ```json
> {
>   ...,
>   "verification": {
>     "email": false
>   }
> }
> ```
> 
> That way the client can read it and show messaging around verifying, while also not blocking the user from accessing parts of the app that do not require verification.

@joshuagraber I suppose the related question is how to determine which endpoints should or should not require verification, either general principles or a line-by-line sense. My personal belief is that the requirement of a JWT and/or Access Token is a pretty good proxy for this, but my personal beliefs mean Jack and Frost, so I'll defer to @josh-chamberlain as to what they think.

**Comment by josh-chamberlain** [comment]
@maxachis @joshuagraber so, my understanding is that we can encode the user's email/github verification status in the token (seems like a good idea to make it readily available), and determine which endpoints require verification however we want. I don't think those ideas are mutually exclusive, unless I am misunderstanding something.

I see the benefits of using access tokens as a proxy, but I also like the idea of frictionless signup and login; that said, anything a user would want to log in for would require them to link github or verify email anyway (having glanced at the API docs) so I think **access tokens as a proxy for requiring email verification is sensible.**

**Comment by maxachis** [comment]
> @maxachis @joshuagraber so, my understanding is that we can encode the user's email/github verification status in the token (seems like a good idea to make it readily available), and determine which endpoints require verification however we want. I don't think those ideas are mutually exclusive, unless I am misunderstanding something.
> 
> I see the benefits of using access tokens as a proxy, but I also like the idea of frictionless signup and login; that said, anything a user would want to log in for would require them to link github or verify email anyway (having glanced at the API docs) so I think **access tokens as a proxy for requiring email verification is sensible.**

As it happens, we do already encode the user's id and/or email (though not their GitHub verification status) in the token, and I do readily use that for various enterprises! 

@josh-chamberlain @joshuagraber 

Anyway, got it! The way this will work is:

1. Users will use a `auth/signup` endpoint to initially sign up and be entered into the database. 
2. They will not be validated until they confirm their email using a token provided in the email, which will need to be submitted to `/auth/validate-email
3. If they try to use `/login` prior to validation, they'll be denied and won't get their access token or JWT or anything else.

**Comment by josh-chamberlain** [comment]
@maxachis this sounds reasonable to me, pending Joshua's eyeballs. For 4, I think removing the user is a bit harsh; we could wait a week, or just keep them as a user and remind them to verify if they try to do stuff a few weeks down the line! We're an "as-needed" service, so I think it's likely that people would sign up, get distracted, and come back much later.

**Comment by joshuagraber** [comment]
Agreed that we probably don't want to automatically remove users. I'd recommend just keeping them around in the un-verified state. Then if they try to log in again, we can issue a new verification token and tell them to go verify.

But the rest of this makes sense to me.

**Comment by maxachis** [comment]
@josh-chamberlain @joshuagraber Alright! I've updated the backend logic to include the new signup and authentication logic, but we're not done! 

1. @joshuagraber Will need to update the frontend so that signups are directed through the new workflow, with `/auth/signup`, `/auth/validate-email` and `/auth/resend-validation-email`
2. @joshuagraber will need to give to _me_ the new address I'll need to provide in the validation email
3. Once that's done and we're confident it works, I'll then need to delete the `/user` `POST` method
4. I'll also need to modify to include the bcc for @josh-chamberlain, which I haven't done partly because we're still doing early testing, and I don't want to spam
5. I may need to update the logic for the simplified creation of test users for @joshuagraber 's sake: Having developer api endpoints, adding test users at the point of database migration, or some other option.

**Comment by joshuagraber** [comment]
> Will need to update the frontend so that signups are directed through the new workflow, with /auth/signup, /auth/validate-email and /auth/resend-validation-email

That's fine, @maxachis. Do you have a branch I can work in so that this all goes into `dev` together?

> will need to give to me the new address I'll need to provide in the validation email

Let's go to `/validate/email/:token`

> I may need to update the logic for the simplified creation of test users for @joshuagraber 's sake: Having developer api endpoints, adding test users at the point of database migration, or some other option.

Yeah, I'd appreciate that. Alternatively, persisting users per DB would make sense to me. So it's not just one users table, but users exist in discrete DB environments and they persist in those environments. That would make things much simpler and much more stable for testing, both manually and with the automated integration testing I'll be adding via Cypress either shortly before or after launch.

**Comment by josh-chamberlain** [comment]
+1 for persistent users where possible/reasonable. This makes intuitive sense to me, especially since we do have the ability to just remove users from the database pretty easily.

**Comment by joshuagraber** [comment]
> just because this would involve the same kind of JWT logic (with period) that we dealt with for the reset-password endpoint!

Oh yeah! Good catch, thank you for remembering that.

**Comment by joshuagraber** [comment]
The client work for this is completed and pushed to `mc_463_create_user_email_verification`. Works like a dream in my testing locally 🎉

**Comment by maxachis** [comment]
@joshuagraber I've add a test user in the stage database that should have a constant email and password. Message me and I can get them your login information!

**Comment by maxachis** [comment]
@josh-chamberlain I've also updated it so that you are included in the bcc. You should have received two signup emails that you were bcc'd on, from two different accounts (one the email you know and love, and one pokemon-themed!)

**Comment by josh-chamberlain** [comment]
@maxachis nice! thank you. I did get those and was _slightly_ confused but figured it was related to testing. Could you please update the template to match the main body of this issue? Also, it should come from `support@` (I don't think `mail@` goes anywhere) and ideally have the appropriate line breaks.

<img width="966" alt="Screen Shot 2024-12-01 at 2 10 08 PM" src="https://github.com/user-attachments/assets/72cff7df-f7e2-4052-87dd-46ed213a0688">

**Comment by maxachis** [comment]
@joshuagraber It seems we've got a bug in the user email verification system at the moment! I tried to sign up, and after clicking on the link in the email, I was redirected to the proper page but got this error:

![image](https://github.com/user-attachments/assets/1f26fabd-0f6b-4d1b-ab62-071595790b0f)

The dreaded spinny-doohickey never stopped spinning, btw.

![image](https://github.com/user-attachments/assets/d23cef5f-8615-412f-ad6a-38747d11a065)
